File: app/desktop/src/services/memory.ts
/**
 * 长期记忆 API 层
 *
 * 数据在后端 memories + memory_tags 表; 搜索按回忆打分排序并强化命中记忆.
 */
import {invoke} from "@tauri-apps/api/core"
import {logger} from "./logger"

/**
 * 一条长期记忆 (与后端 MemoryRecord camelCase 对齐)
 */
export interface MemoryRecord {
	id: number
	content: string
	type: string
	importance: number
	confidence: number
	tags: string[]
	accessCount: number
	lastAccessedAt: number | null
	expiresAt: number | null
	status: string
	createdAt: number
	updatedAt: number
	recallScore?: number
}

/**
 * 记忆写入参数
 */
export interface MemoryInput {
	content: string
	type: string
	importance: number
	confidence: number
	tags: string[]
	expiresAt?: number | null
}

/**
 * 全部记忆 (按创建时间倒序)
 */
export const listMemories = async (limit = 200): Promise<MemoryRecord[]> => {
	try {
		return await invoke<MemoryRecord[]>("memory_list", {limit})
	} catch (error) {
		await logger.error("[memory] 获取记忆列表失败", error)
		return []
	}
}

/**
 * 搜索记忆 (内容 / 标签命中, 按回忆打分排序)
 */
export const searchMemories = async (query: string, limit = 20): Promise<MemoryRecord[]> => {
	try {
		return await invoke<MemoryRecord[]>("memory_search", {args: {query, limit}})
	} catch (error) {
		await logger.error("[memory] 搜索记忆失败", error)
		return []
	}
}

/**
 * 新建记忆
 */
export const createMemory = async (input: MemoryInput): Promise<MemoryRecord | null> => {
	try {
		return await invoke<MemoryRecord>("memory_create", {args: input})
	} catch (error) {
		await logger.error("[memory] 创建记忆失败", error)
		return null
	}
}

/**
 * 更新记忆
 */
export const updateMemory = async (id: number, input: MemoryInput): Promise<MemoryRecord | null> => {
	try {
		return await invoke<MemoryRecord>("memory_update", {id, args: input})
	} catch (error) {
		await logger.error("[memory] 更新记忆失败", error)
		return null
	}
}

/**
 * 删除记忆
 */
export const deleteMemory = async (id: number): Promise<boolean> => {
	try {
		await invoke("memory_delete", {id})
		return true
	} catch (error) {
		await logger.error("[memory] 删除记忆失败", error)
		return false
	}
}

File: app/desktop/src-tauri/src/commands/memory.rs
//! 长期记忆命令模块
//!
//! memories + memory_tags 表, 提供列表 / 搜索(回忆打分) / 创建 / 更新 / 删除.

use crate::db;
use crate::memory::model::{MemoryInput, MemoryRecord};
use crate::memory::repository;

fn now() -> i64 {
	std::time::SystemTime::now()
		.duration_since(std::time::UNIX_EPOCH)
		.map(|d| d.as_secs() as i64)
		.unwrap_or(0)
}


/// 全部记忆 (按创建时间倒序)
/// invoke("memory_list", { args: { limit: 200 } })
#[tauri::command]
pub fn memory_list(
	state: tauri::State<'_, db::Db>,
	limit: Option<u32>,
) -> Result<Vec<MemoryRecord>, String> {
	let conn = state
		.0
		.lock()
		.map_err(|e| format!("获取数据库连接失败: {e}"))?;
	repository::list(&conn, limit.unwrap_or(200).clamp(1, 1000) as usize, now())
}

/// 搜索记忆 (内容 / 标签命中, 按回忆打分排序; 命中记忆访问次数 +1 强化)
/// invoke("memory_search", { args: { query: "项目", limit: 20 } })
#[tauri::command]
pub fn memory_search(
	state: tauri::State<'_, db::Db>,
	args: MemorySearchArgs,
) -> Result<Vec<MemoryRecord>, String> {
	let conn = state
		.0
		.lock()
		.map_err(|e| format!("获取数据库连接失败: {e}"))?;
	let limit = args.limit.unwrap_or(20).clamp(1, 200) as usize;
	let query = args.query.trim();
	repository::search_scored(&conn, query, limit, now())
}

/// 新建记忆
/// invoke("memory_create", { args: { content, type, importance, confidence, tags, expiresAt? } })
#[tauri::command]
pub fn memory_create(
	state: tauri::State<'_, db::Db>,
	args: MemoryInput,
) -> Result<MemoryRecord, String> {
	let input = args.normalize()?;
	let conn = state
		.0
		.lock()
		.map_err(|e| format!("获取数据库连接失败: {e}"))?;
	let id = repository::create(&conn, &input, now())?;
	repository::get(&conn, id)?.ok_or_else(|| "记忆创建失败".to_string())
}

/// 更新记忆
/// invoke("memory_update", { id, args: { ... } })
#[tauri::command]
pub fn memory_update(
	state: tauri::State<'_, db::Db>,
	id: i64,
	args: MemoryInput,
) -> Result<MemoryRecord, String> {
	let input = args.normalize()?;
	let conn = state
		.0
		.lock()
		.map_err(|e| format!("获取数据库连接失败: {e}"))?;
	repository::update(&conn, id, &input, now())?;
	repository::get(&conn, id)?.ok_or_else(|| "记忆不存在".to_string())
}

/// 删除记忆
/// invoke("memory_delete", { id })
#[tauri::command]
pub fn memory_delete(state: tauri::State<'_, db::Db>, id: i64) -> Result<(), String> {
	let conn = state
		.0
		.lock()
		.map_err(|e| format!("获取数据库连接失败: {e}"))?;
	repository::delete(&conn, id)
}

/// 搜索参数
#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemorySearchArgs {
	pub query: String,
	#[serde(default)]
	pub limit: Option<u32>,
}
